networking.bonds: add support for arbitrary driverOptions

Profpatsch opened the pull request:
Until now the four attributes available very selectively provided a small
subset, while copying upstream documentation.

We make driver options an arbitrary key-value set and point to kernel
documentation, which is always up-to-date. This way every option can be set.
The four already existing options are deprecated with a warning.

###### Motivation for this change


###### Things done

- [x] Tested using sandboxing
- Built on platform(s)
   - [x] NixOS
- [x] ran & corrected vm-test

Profpatsch committed:
networking.bonds: add support for arbitrary driverOptions

Until now the four attributes available very selectively provided a small
subset, while copying upstream documentation.

We make driver options an arbitrary key-value set and point to kernel
documentation, which is always up-to-date. This way every option can be set.
The four already existing options are deprecated with a warning.

aszlig commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -550,11 +550,28 @@ in
+            description = ''
+              Options for the bonding driver.
+              Documentation can be found in
+              https://www.kernel.org/doc/Documentation/networking/bonding.txt
```
`<link xlink:href="https://www.kernel.org/doc/Documentation/networking/bonding.txt" />`

Profpatsch replied:
fix’d

aszlig commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -565,6 +582,7 @@ in
             example = 100;
             type = types.nullOr types.int;
             description = ''
+              DEPRECATED, use `driverOptions`.
```
Instead of writing "DEPRECATED" in the description, I'd either pass down miimon/lacp_rate/mode/... to `driverOptions` in the definition or properly deprecate/remove it via [`rename.nix`](https://github.com/NixOS/nixpkgs/blob/9f334c2732e4d8606ccb09b14893aad5c37bb9da/nixos/modules/rename.nix).

Profpatsch replied:
Hm, it’s not really renamed, it’s been deprecated and is still being merged right now (with a warning), but will be removed in 17.09 or something.

fpletz replied:
This is not true, at least not using the scripted networking. This breaks bonding for me. Where do you merge the deprecated settings?

fpletz replied:
So we have a LACP bond that is configured with the deprecated options but after this change came up with balance-rr instead. Which broke its networking. :disappointed:

This should either trigger an error or merging should work. As this breaks peoples' setups I'm inclined to revert this for now.

:+1: on the change, though.

fpletz replied:
Okay, after trying to reproduce this: This problem only applies to networkd. Investigating.

aszlig commented:
I'm not sure why you deprecate the explicit options instead of the generic `driverOptions`, because they can work happily side-by-side by just passing those options to `driverOptions` by default, which also has the advantage that for eg. `miimon` an integer can be passed instead of a string, so it's even more type-safe.

Profpatsch commented:
> why you deprecate the explicit options

Because they have to stay in sync with upstream options. And who watches that?
As for the extra type-safety, I’m not sure that’s a very big gain compared to the possible mismatch after one/two years or so.

aszlig commented on nixos/modules/tasks/network-interfaces-scripted.nix:
```diff
@@ -288,10 +301,11 @@ in
-                ${optionalString (v.miimon != null) "miimon ${toString v.miimon}"} \
-                ${optionalString (v.xmit_hash_policy != null) "xmit_hash_policy ${toString v.xmit_hash_policy}"} \
-                ${optionalString (v.lacp_rate != null) "lacp_rate ${toString v.lacp_rate}"}
+              ${let opts = (mapAttrs (const toString)
```
Weird indentation... I'd join this into something like: `ip link add name "${n}" type bond ${let ...`

Profpatsch replied:
The indentation is done by the function that creates the lines.

```
ip link add … \
  mode foo \
  other_option bar
```

aszlig replied:
I meant the indentation on the Nix side.

Profpatsch replied:
Yeah, it’s like that because otherwise the first line will be indented by two spaces more (since it starts with two spaces already).

Profpatsch committed:
networking.networkd: adjust autmatic mapping of bonds

Since the bonds interface changed to a lot more possible values we create a
mapping of kernel bond attribute names and values to networkd attributes.
Those match for the most part, but have to transformed slightly.

There is also an assert that unknown options won’t slip through silently.

globin commented:
Please don't merge stuff like this yourself which can break (and broke) networking and can possibly lock people out of machines that are not on-site. This heavily frustrates people and shipping broken networking code even on unstable is highly unprofessional..

Profpatsch commented:
Point taken; the code itself is correct, the interaction of automatic translation to networkd options left out the deprecated options.
If the deprecated options are not used, it works in any case, as well as when networkd is not used.

Profpatsch commented:
@qknight noticed me that the descriptions are off. I’ll fix.

Profpatsch commented:
Fixed in https://github.com/NixOS/nixpkgs/pull/23555
